Forward declaration suggested for fully-defined struct

For this header:

```c
#define MACRO(ptr) ((*ptr) + 0)

struct foo {
	int x;
};

static inline void foo(struct foo *foo) {
	MACRO(&foo);
}
```

IWYU suggests adding a `struct foo;` forward declaration, despite the fact that `foo.h` already defines the struct:

```console
tavianator@graphene $ include-what-you-use foo.h

foo.h should add these lines:
struct foo;

foo.h should remove these lines:

The full include-list for foo.h:
struct foo;
---
tavianator@graphene $ include-what-you-use --version
include-what-you-use 0.21 based on clang version 17.0.6
```

<details>
<summary>Verbose logs</summary>

```console
tavianator@graphene $ include-what-you-use foo.h -Xiwyu -v7
Setting verbose-level to 7
Search path: /usr/lib/clang/17/include/ (system)
Search path: /usr/local/include/ (system)
Search path: /usr/include/ (system)
Found include: <built-in>:-1 -> foo.h
Added to main compilation unit: foo.h
[ Exiting to  ] foo.h:1:1
[ #define     ] foo.h:1:9: MACRO
[ Use macro   ] foo.h:8:2: MACRO (from foo.h:1:9)
Marked full-info use of macro MACRO (from foo.h) at foo.h:8:2
Intends-to-provide for foo.h:
   foo.h
foo.h:3:8: (1) [ RecordDecl ] 0x55c509af8250 struct foo {
}
foo.h:4:6: (2) [ FieldDecl ] 0x55c509af8308 int x
foo.h:4:2: (3) [ BuiltinTypeLoc ] int
foo.h:7:20: (1) [ FunctionDecl ] 0x55c509af8508 static inline void foo(struct foo *foo)
foo.h:7:15: (2) [ FunctionProtoTypeLoc ] void (struct foo *)
foo.h:7:15: (3) [ BuiltinTypeLoc ] void
foo.h:7:36: (3) [ ParmVarDecl ] 0x55c509af8410 struct foo *foo
foo.h:7:24: (4) [ PointerTypeLoc ] struct foo *
foo.h:7:24: (5) [ ElaboratedTypeLoc ] struct foo
foo.h:7:31: (6) [ RecordTypeLoc ] struct foo
Marked fwd-decl use of decl 0x55c509af8250 foo (from foo.h:3:8) at foo.h:7:31
foo.h:7:41: (2) [ CompoundStmt ] 0x55c509af8798 CompoundStmt 0x55c509af8798
`-ParenExpr 0x55c509af8778 'struct foo *'
  `-BinaryOperator 0x55c509af8758 'struct foo *' '+'
    |-ImplicitCastExpr 0x55c509af86e8 'struct foo *' <LValueToRValue>
    | `-ParenExpr 0x55c509af86a8 'struct foo *' lvalue
    |   `-UnaryOperator 0x55c509af8690 'struct foo *' lvalue prefix '*' cannot overflow
    |     `-UnaryOperator 0x55c509af8678 'struct foo **' prefix '&' cannot overflow
    |       `-DeclRefExpr 0x55c509af8600 'struct foo *' lvalue ParmVar 0x55c509af8410 'foo' 'struct foo *'
    `-IntegerLiteral 0x55c509af86c8 'int' 0

foo.h:8:2 <Spelling=foo.h:1:20>: (3) [ ParenExpr ] 0x55c509af8778 ParenExpr 0x55c509af8778 'struct foo *'
`-BinaryOperator 0x55c509af8758 'struct foo *' '+'
  |-ImplicitCastExpr 0x55c509af86e8 'struct foo *' <LValueToRValue>
  | `-ParenExpr 0x55c509af86a8 'struct foo *' lvalue
  |   `-UnaryOperator 0x55c509af8690 'struct foo *' lvalue prefix '*' cannot overflow
  |     `-UnaryOperator 0x55c509af8678 'struct foo **' prefix '&' cannot overflow
  |       `-DeclRefExpr 0x55c509af8600 'struct foo *' lvalue ParmVar 0x55c509af8410 'foo' 'struct foo *'
  `-IntegerLiteral 0x55c509af86c8 'int' 0

foo.h:8:2 <Spelling=foo.h:1:28>: (4) [ BinaryOperator ] 0x55c509af8758 BinaryOperator 0x55c509af8758 'struct foo *' '+'
|-ImplicitCastExpr 0x55c509af86e8 'struct foo *' <LValueToRValue>
| `-ParenExpr 0x55c509af86a8 'struct foo *' lvalue
|   `-UnaryOperator 0x55c509af8690 'struct foo *' lvalue prefix '*' cannot overflow
|     `-UnaryOperator 0x55c509af8678 'struct foo **' prefix '&' cannot overflow
|       `-DeclRefExpr 0x55c509af8600 'struct foo *' lvalue ParmVar 0x55c509af8410 'foo' 'struct foo *'
`-IntegerLiteral 0x55c509af86c8 'int' 0

(For type struct foo):
Trying to determine use location for 'struct foo {
}'
Macro is defined in 'foo.h'
Attributing use of 'struct foo {
}' to spelling location at foo.h:1:28
Marked full-info use of decl 0x55c509af8250 foo (from foo.h:3:8) at foo.h:1:28
foo.h:8:2 <Spelling=foo.h:1:21>: (5) [ ImplicitCastExpr ] 0x55c509af86e8 ImplicitCastExpr 0x55c509af86e8 'struct foo *' <LValueToRValue>
`-ParenExpr 0x55c509af86a8 'struct foo *' lvalue
  `-UnaryOperator 0x55c509af8690 'struct foo *' lvalue prefix '*' cannot overflow
    `-UnaryOperator 0x55c509af8678 'struct foo **' prefix '&' cannot overflow
      `-DeclRefExpr 0x55c509af8600 'struct foo *' lvalue ParmVar 0x55c509af8410 'foo' 'struct foo *'

foo.h:8:2 <Spelling=foo.h:1:21>: (6) [ ParenExpr ] 0x55c509af86a8 ParenExpr 0x55c509af86a8 'struct foo *' lvalue
`-UnaryOperator 0x55c509af8690 'struct foo *' lvalue prefix '*' cannot overflow
  `-UnaryOperator 0x55c509af8678 'struct foo **' prefix '&' cannot overflow
    `-DeclRefExpr 0x55c509af8600 'struct foo *' lvalue ParmVar 0x55c509af8410 'foo' 'struct foo *'

foo.h:8:2 <Spelling=foo.h:8:8>: (7) [ UnaryOperator ] 0x55c509af8690 UnaryOperator 0x55c509af8690 'struct foo *' lvalue prefix '*' cannot overflow
`-UnaryOperator 0x55c509af8678 'struct foo **' prefix '&' cannot overflow
  `-DeclRefExpr 0x55c509af8600 'struct foo *' lvalue ParmVar 0x55c509af8410 'foo' 'struct foo *'

foo.h:8:2 <Spelling=foo.h:8:9>: (8) [ UnaryOperator ] 0x55c509af8678 UnaryOperator 0x55c509af8678 'struct foo **' prefix '&' cannot overflow
`-DeclRefExpr 0x55c509af8600 'struct foo *' lvalue ParmVar 0x55c509af8410 'foo' 'struct foo *'

foo.h:8:2 <Spelling=foo.h:8:9>: (9) [ DeclRefExpr ] 0x55c509af8600 DeclRefExpr 0x55c509af8600 'struct foo *' lvalue ParmVar 0x55c509af8410 'foo' 'struct foo *'

Trying to determine use location for 'struct foo *foo'
Macro is defined in 'foo.h'
Attributing use of 'struct foo *foo' to spelling location at foo.h:8:9
Marked full-info use of decl 0x55c509af8410 foo (from foo.h:7:36) at foo.h:8:9
foo.h:8:2 <Spelling=foo.h:1:30>: (5) [ IntegerLiteral ] 0x55c509af86c8 IntegerLiteral 0x55c509af86c8 'int' 0

--- Calculating IWYU violations for foo.h ---
Ignoring fwd-decl use of foo (foo.h:7:31): dfn is present: foo.h:3:8
Moving foo from full use to fwd-decl: definition found later in file (foo.h:1:28)
Ignoring full use of foo (foo.h:8:9): declared inside a function
Ignoring symbol use of MACRO (foo.h:8:2): defined in same file
foo.h:1:28: warning: foo needs a declaration, but does not provide or directly #include one.

foo.h should add these lines:
struct foo;

foo.h should remove these lines:

The full include-list for foo.h:
struct foo;
---
```

</details>

--- a/iwyu.cc
+++ b/iwyu.cc
@@ -2017,10 +2017,14 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
       return true;
 
     for (const Stmt* child : expr->children()) {
-      const Type* type = GetTypeOf(cast<Expr>(child));
+      // Dig past implicit casts and parens to get the real location.
+      const Expr* term = cast<Expr>(child)->IgnoreImpCasts()->IgnoreParens();
+
+      // If the term expression is pointer-typed we need the full types for
+      // pointer arithmetic.
+      const Type* type = GetTypeOf(term);
       if (type->isPointerType()) {
-        // It's a pointer-typed expression.
-        ReportTypeUse(CurrentLoc(), type, DerefKind::RemoveRefsAndPtr);
+        ReportTypeUse(GetLocation(term), type, DerefKind::RemoveRefsAndPtr);
       }
     }
 

--- a/tests/cxx/pointer_arith.cc
+++ b/tests/cxx/pointer_arith.cc
@@ -103,6 +103,34 @@ void BuiltinPointerArithmetic() {
   x = pi - &i;
 }
 
+// Pointer arithmetic is inside a macro, but only using arguments, macro doesn't
+// know about the types involved. Should be reported at expansion.
+#define DISTANCE(ptr1, ptr2) ((ptr1) - (ptr2))
+
+void MacroArgPointerArithmetic() {
+  // IWYU: IndirectClass needs a declaration
+  IndirectClass* p1 = &ic1;
+  // IWYU: IndirectClass needs a declaration
+  IndirectClass* p2 = &ic2;
+
+  // IWYU: IndirectClass is...*indirect.h
+  (void)DISTANCE(p2, p1);
+}
+
+// Pointer arithmetic is entirely inside macro, caller doesn't know about the
+// types involved. Should be reported at spelling inside macro.
+
+// IWYU: IndirectClass needs a declaration
+static IndirectClass* table;
+
+// IWYU: IndirectClass is...*indirect.h
+#define GET_NTH(n) (*(table + n))
+
+void MacroBodyPointerArithmetic() {
+  // Use is entirely contained in macro.
+  (void)GET_NTH(10);
+}
+
 /**** IWYU_SUMMARY
 
 tests/cxx/pointer_arith.cc should add these lines:

